File: golang/container/bit.go
package container

// 树状数组
// 数组中的元素储存在[1, n]
type BinaryIndexedTree struct {
	n    int
	data []int
}

func InitBinaryIndexedTree(n int) *BinaryIndexedTree {
	bit := &BinaryIndexedTree{}
	bit.n = n
	bit.data = make([]int, n+1)
	return bit
}

// 统计前n个数之和 (n从1开始)
func (x *BinaryIndexedTree) Sum(n int) int {
	sum := 0
	for n > 0 {
		sum += x.data[n]
		n = n & (n - 1)
	}
	return sum
}

// 给第i个元素+a (i从0开始)
func (x *BinaryIndexedTree) Add(i, a int) {
	// 为了方便起见，i从0开始
	n := i + 1
	for n <= x.n {
		x.data[n] += a
		n += n & -n
	}
}

File: python/container/bit.py
"""
树状数组: 统计前n个元素和
元素储存在 [1, n]
"""


class BinaryIndexedTree:
    def __init__(self, n: int):
        self.n = n
        self.data = [0] * (n + 1)

    def sum(self, i: int) -> int:
        """
        统计 [0, i] 元素的和
        i 范围 [0, n)
        """
        n = i + 1
        total = 0
        while n:
            total += self.data[n]
            n = n & (n - 1)
        return total

    def add(self, i: int, a: int) -> None:
        """
        为索引 i 的元素**增加a**
        i 范围 [0, n)
        """
        n = i + 1
        while n <= self.n:
            self.data[n] += a
            n += n & -n


def main():
    bit = BinaryIndexedTree(5)
    bit.add(0, 5)
    bit.add(1, 3)
    bit.add(2, 7)
    bit.add(3, 2)
    bit.add(4, 8)

    print(bit.sum(0))  # 5
    print(bit.sum(1))  # 8
    print(bit.sum(2))  # 15
    print(bit.sum(3))  # 17
    print(bit.sum(4))  # 25

    # update 0
    bit.add(0, -1)

    print(bit.sum(0))  # 4
    print(bit.sum(1))  # 7
    print(bit.sum(2))  # 14


main()
